client/transaction-pool/src/lib.rs

```diff
@@ -186,14 +175,27 @@ impl<PoolApi, Block> BasicPool<PoolApi, Block>
 		)
 	}
 
+	/// Create new basic transaction pool for a light node with the provided api.
+	pub fn new_light(
+		options: sc_transaction_graph::Options,
+		pool_api: Arc<PoolApi>,
+		prometheus: Option<&PrometheusRegistry>,
+		spawner: impl SpawnNamed,
+	) -> Self {
```

Comment:
I think one of the goal would have been also to reverse how the node is built.

Right now we do:
1. we start with new_light/new_full
2. then we add stuff on top of it

It would be best if we could:
1. build the node components stuff
2. then call build_light/build_full/(build_ops) at the end

I'm not sure how to implement this but if it happens it means we could have 1 single function in bin/node-template/src/service.rs

Reply:
(wrong place to put this comment)

Reply:
Sounds good and much easier to understand!

Reply:
>It would be best if we could:
> * build the node components stuff
> * then call build_light/build_full/(build_ops) at the end

That could possibly work! The way I can see it would involve setting up a builder with a lot of generics, then calling `build_light`/`build_full` on that:

```rust
let builder = sc_service::create_builder<
    // Base stuff
    Block, RuntimeApi, Executor
    // Chain selection
    LongestChain,
    // Import queue
    AuraImportQueue<GrandpaBlockImport>,
    GrandpaFinalityProofProvider,
    ...
>();

builder.build_full();
```

Reply:
I will test out this idea. Ideally, the generics wouldn't be any more complicated than that.

Reply:
See https://github.com/paritytech/substrate/pull/6600.

Reply:
The changes in this PR need to get merged in the first place to implement this.